### DISCUSSION TITLE
Parametrize Normal Distribution with Precision Matrix

### DISCUSSION THREAD

**Discussion by alexpeters1208** [cornellius-gp/gpytorch#1999] [Q&A] [answered]
Hello all. I'm working on a kernel wrapper that implements some neighbor-based approximations for speeding up GP computations. These approximations yield a sparse precision matrix, and I would like to be able to use that directly in my GPyTorch model. However, at the moment, I cannot figure out how to parametrize the GPyTorch MVN with its precision matrix. I know that PyTorch's MVN offers a precision matrix parametrization, but cannot find it in GPyTorch. If there is no such implementation, any clues as to how I might go about making one?

Thank you!

**Comment by gpleiss** [comment, marked as answer]
GPyTorch's MVN parameterizes the covariance matrix via a "LazyTensor" (https://docs.gpytorch.ai/en/stable/lazy.html), which essentially corresponds to linear operator classes found in other packages like SciPy. All operations on the covariance matrix generally tend to avoid directly instantiating the covariance matrix, and instead access it through matrix multiplication or solves.

The way to implement this in GPyTorch would be to implement a `SparsePrecisionLazyTensor`, which would be parameterized by a sparse precision matrix. You would then want to override the `inv_matmul`, `inv_quad`, and `inv_quad_logdet` to essentially perform matrix multiplication with the precision matrix.

I do know there are some efforts currently underway to add nearest neighbor (i.e. sparse precision) approximations to GPyTorch, so hopefully these will be features in the library in the not to distant future.

**Reply by alexpeters1208** [reply]
Thanks Geoff, this was very helpful! Is there a ticket for that feature that I could keep up with? I will start working on it either way. I appreciate your input.

**Reply by gpleiss** [reply]
No ticket at the moment.
